README
NAME

    WebService::HipChat

VERSION

    version 0.0601

SYNOPSIS

        my $hc = WebService::HipChat->new(auth_token => 'abc');
        $hc->send_notification('Room42', { color => 'green', message => 'allo' });
    
        # get paged results:
        my $res = $hc->get_emoticons;
        my @emoticons = @{ $res->{items} };
        while ($res = $hc->next($res)) {
            push @emoticons, @{ $res->{items} };
        }

DESCRIPTION

    This module provides bindings for the HipChat API v2
    <https://www.hipchat.com/docs/apiv2>. It also provides the command line
    utility hipchat-send.

METHODS

    All methods return a hashref. The $room param can be the id or name of
    the room. The $user param can be the id, email address, or @mention
    name of the user. If a resource does not exist for the given
    parameters, undef is returned.

 get_rooms

        get_rooms()
        get_rooms(query => { 'start-index' => 0, 'max-results' => 100 });

    Example response:

        {
          items => [
            {
              id => 2,
              links => {
                self => "https://hipchat.com/v2/room/2",
                webhooks => "https://hipchat.com/v2/room/2/webhook",
              },
              name => "General Discussion",
            },
            {
              id => 3,
              links => {
                self => "https://hipchat.com/v2/room/3",
                webhooks => "https://hipchat.com/v2/room/3/webhook",
              },
              name => "Important Stuff",
            },
          links => { self => "https://hipchat.com/v2/room" },
          maxResults => 100,
          startIndex => 0,
        }

 get_room

        get_room($room)

    Example response:

        {
          created => "2014-06-25T02:28:04",
          guest_access_url => undef,
          id => 2,
          is_archived => 0,
          is_guest_accessible 0,
          last_active => "2014-07-19T02:40:55+00:00",
          links => {
            self => "https://hipchat.com/v2/room/2",
            webhooks => "https://hipchat.com/v2/room/2/webhook",
          },
          name => "General Discussion",
          owner => {
            id => 1,
            links => { self => "https://hipchat.com/v2/user/1" },
            mention_name => "bob",
            name => "Bob Williams",
          },
          participants => [],
          privacy => "public",
          statistics => {
            links => { self => "https://hipchat.com/v2/room/2/statistics" },
          },
          topic => "hipchat commands",
          xmpp_jid => "1_general_discussion\@conf.btf.hipchat.com",
        }

 create_room

        create_room({ name => 'monkeys' })

    Example response:

        {
          id => 46,
          links => { self => "https://hipchat.com/v2/room/46" },
        }

 update_room

        update_room($room, {
            is_archived         => JSON::false,
            is_guest_accessible => JSON::false,
            name                => "Jokes",
            owner               => { id => 17 },
            privacy             => "public",
            topic               => "funny jokes",
        });

 set_topic

        set_topic($room, 'new topic');

 delete_room

        delete_room($room)

 send_notification

        send_notification($room, { color => 'green', message => 'allo' });

 get_webhooks

        get_webhooks($room)
        get_webhooks($room, query => { 'start-index' => 0, 'max-results' => 100 });

    Example response:

        {
          items => [
            {
              event => "room_message",
              id => 1,
              links => { self => "https://hipchat.com/v2/room/API/webhook/1" },
              name => "hook1",
              pattern => undef,
              url => "http://yourdomain.org/hipchat-webhook",
            },
          ],
          links => { self => "https://hipchat.com/v2/room/API/webhook" },
          maxResults => 100,
          startIndex => 0,
        }

 create_webhook

        create_webhook($room, {
            url   => 'http://yourdomain.org/hipchat-webhook'
            event => 'room_message',
            name  => 'hook1',
        });

 send_private_msg

        send_private_msg($user, { message => 'allo' });

 get_members

        get_members($room);
        get_members($room, query => { 'start-index' => 0, 'max-results' => 100 });

    Example response:

        {
          items => [
            {
              id => 73,
              links => { self => "https://hipchat.com/v2/user/73" },
              mention_name => "momma",
              name => "Yo Momma",
            },
            {
              id => 23,
              links => { self => "https://hipchat.com/v2/user/23" },
              mention_name => "jackie",
              name => "Jackie Chan",
            },
          ],
          links => { self => "https://hipchat.com/v2/room/Test/member" },
          maxResults => 100,
          startIndex => 0,
        }

 add_member

    Adds a user to a room.

        add_member($room, $user);

 remove_member

    Removes a user from a room.

        remove_member($room, $user);

 get_users

        get_users()
        get_users(query => { 'start-index' => 0, 'max-results' => 100 });

    Example response:

        {
          items => [
            {
              id => 1,
              links => { self => "https://hipchat.com/v2/user/1" },
              mention_name => "magoo",
              name => "Matt Wondercookie",
            },
            {
              id => 3,
              links => { self => "https://hipchat.com/v2/user/3" },
              mention_name => "racer",
              name => "Brian Wilson",
            },
          ],
          links => { self => "https://hipchat.com/v2/user" },
          maxResults => 100,
          startIndex => 0,
        }

 get_user

        get_user($user)

    Example response:

        {
          created        => "2014-06-20T03:00:28",
          email          => 'matt@foo.com',
          group          => {
                              id => 1,
                              links => { self => "https://hipchat.com/v2/group/1" },
                              name => "Everyone",
                            },
          id             => 1,
          is_deleted     => 0,
          is_group_admin => 1,
          is_guest       => 0,
          last_active    => 1405718128,
          links          => { self => "https://hipchat.com/v2/user/1" },
          mention_name   => "magoo",
          name           => "Matt Wondercookie",
          photo_url      => "https://hipchat.com/files/photos/1/abc.jpg",
          presence       => {
                              client => {
                                type => "http://hipchat.com/client/linux",
                                version => 98,
                              },
                              idle => 3853,
                              is_online => 1,
                              show => "away",
                            },
          timezone       => "America/New_York",
          title          => "Hacker",
          xmpp_jid       => '1_1@chat.hipchat.com',
        }

 get_emoticons

        get_emoticons()
        get_emoticons(query => { 'start-index' => 0, 'max-results' => 100 });

    Example response:

        {
          items => [
            {
              id => 166,
              links => { self => "https://hipchat.com/v2/emoticon/166" },
              shortcut => "dog",
              url => "https://hipchat.com/files/img/emoticons/1/dog.png",
            },
          ],
          links => { self => "https://hipchat.com/v2/emoticon" },
          maxResults => 100,
          startIndex => 0,
        }

 get_emoticon

        get_emoticon()

    Example response:

        {
          creator => {
            id => 11,
            links => { self => "https://hipchat.com/v2/user/11" },
            mention_name => "bob",
            name => "Bob Ray",
          },
          height => 30,
          id => 203,
          links => { self => "https://hipchat.com/v2/emoticon/203" },
          shortcut => "dog",
          url => "https://hipchat.com/files/img/emoticons/1/dog.png",
          width => 30,
        }

 get_room_history

        $hc->get_room_history($room)
        $hc->get_room_history($room, { 'max-results' => 5 });

    Example response:

       {
        items        [
            [0] {
                date       "2014-11-13T10:48:33.322506+00:00",
                from       {
                    id             123456,
                    links          {
                        self   "https://api.hipchat.com/v2/user/123456"
                    },
                    mention_name   "Bob",
                    name           "Bob Williams"
                },
                id         "38988c8c-9120-44ce-87c5-6731a7a3b6",
                mentions   [],
                message    "heres a message and a link http://www.sun.com/",
                type       "message"
            },
            [1] {
                date       "2014-11-13T10:49:02.377436+00:00",
                from       {
                    id             123456,
                    links          {
                        self   "https://api.hipchat.com/v2/user/123456"
                    },
                    mention_name   "Bob",
                    name           "Bob Williams"
                },
                id         "c1f47537-6506-4f46-b820-eaade5adc5",
                mentions   [],
                message    "A message",
                type       "message"
            }
        ],
        links        {
            self   "https://api.hipchat.com/v2/room/XXX/history/latest"
        },
        maxResults   2,
        startIndex   0
       }

 share_link

        $hc->share_link($room, { message => 'msg', link => 'http://www.sun.com' });

 share_file

        $hc->share_file($room, { message => 'msg', file => '/tmp/file.png' });

 next

        next($data)

    Returns the next page of data for paginated responses.

    Example:

        my $res = $hc->get_emoticons;
        my @emoticons = @{ $res->{items} };
        while ($res = $hc->next($res)) {
            push @emoticons, @{ $res->{items} };
        }

CONTRIBUTORS

    Chris Hughes <chris@lokku.com>

AUTHOR

    Naveed Massjouni <naveed@vt.edu>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Naveed Massjouni.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
==================================================
Changes from 2004-12-04 00:00:00 +0000 to present.
==================================================

-------------------------------------------
version 0.0601 at 2014-12-02 07:36:31 +0000
-------------------------------------------

  Change: 74119649e5c83b443750f99e1451f4a21320959e
  Author: Naveed Massjouni <naveed@vt.edu>
  Date : 2014-12-02 02:35:33 +0000

    fix share_file() issue by requiring latest WebService::Client 

-------------------------------------------
version 0.0600 at 2014-11-21 19:07:03 +0000
-------------------------------------------

  Change: d932b4437663b4ae22418c4210b23214ea9cc8d2
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-21 14:07:03 +0000

    v0.0600 added Chris Hughes to CONTRIBUTORS 

  Change: 78bca094696c62f0a9dc5ce85dd0d99072c701f5
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-21 13:56:56 +0000

    Merge pull request #6 from chrisspang/master

    Use WebService::Client new features to build request 

  Change: f0624ec9f3b9273a08d4f7b48a186658d8e9bb78
  Author: Chris Hughes <chris@lokku.com>
  Date : 2014-11-20 13:45:51 +0000

    Use WebService::Client new features to build request 

  Change: 41981e629a5be2cd752c3d9e203829e29d16ddf8
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 14:30:02 +0000

    Merge pull request #4 from chrisspang/master

    Add share_file() 

  Change: 196cc1d3d2d7b3d8cf5bd196fd2d66283a146ceb
  Author: Chris Hughes <chris@lokku.com>
  Date : 2014-11-14 12:26:44 +0000

    Add share_file() 

-------------------------------------------
version 0.0500 at 2014-11-13 16:25:34 +0000
-------------------------------------------

  Change: 5714e0c68a9b099906d82cf6fa1e709944f7bc2a
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-13 11:25:34 +0000

    v0.0500 

  Change: 6ee488da01b74a050f062fc43d14655bafe7f0f9
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-13 11:23:41 +0000

    Merge pull request #3 from chrisspang/master

    Switch to WebService::Client 

  Change: ba31f78d963ac19d0e1b07f66379dfc4f22c23de
  Author: Chris Hughes <chris@lokku.com>
  Date : 2014-11-13 11:09:20 +0000

    Added share_link() get_room_history() 

  Change: 1901fb36d133f1de4496fa4745b5121d2a344db9
  Author: Chris Hughes <chris@lokku.com>
  Date : 2014-11-13 10:34:14 +0000

    Switch to WebService::Client 

-------------------------------------------
version 0.0402 at 2014-11-13 08:01:30 +0000
-------------------------------------------

  Change: 1030abbb8feb155cf869e91a3523f98bbe0af56d
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-13 03:01:30 +0000

    v0.0402 switched to WebService::Client 

  Change: 26ec34eb1dbcd21728bd57adbe5663db4adf0e11
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-12 12:26:15 +0000

    Merge pull request #1 from chrisspang/master

    Simple usage documentation fix 

  Change: 439e594546ebcc1ba89b5778ce48a2d3861fc849
  Author: Chris Hughes <chris@lokku.com>
  Date : 2014-11-12 17:05:33 +0000

    Simple usage documentation fix 

  Change: aefc5c74169b7169164bcfe5638876becaf6818a
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-01 14:56:19 +0000

    fixed get_users documentation 

-------------------------------------------
version 0.0400 at 2014-11-01 10:13:18 +0000
-------------------------------------------

  Change: 7d646a3a71b529838b8db506e9636cb692c1ece0
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-01 06:13:18 +0000

    v0.0400 added support for providing query params for getting
    collections 

-------------------------------------------
version 0.0301 at 2014-10-28 23:44:17 +0000
-------------------------------------------

  Change: ce1f7326aaf0309f4c7ec3615df12b088043c9a8
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-10-28 19:44:17 +0000

    v0.0301 fixed pagination 

-------------------------------------------
version 0.0300 at 2014-10-28 23:29:21 +0000
-------------------------------------------

  Change: 1aebb704929275ee436bdacf32716bce65bf4baf
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-10-28 19:29:21 +0000

    added support for pagination via next() 

-------------------------------------------
version 0.0201 at 2014-07-26 09:06:34 +0000
-------------------------------------------

  Change: 48058a6280676bf27d7d4c73df11ac5326bb8d4d
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-26 05:06:34 +0000

    v0.0201 updated long options to hipchat-send 

  Change: 25521b26fc2458352297cebf23e425f766535654
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-26 04:03:02 +0000

    v0.0200 added hipchat-send command line utility for sending messages 

-------------------------------------------
version 0.0100 at 2014-07-21 08:58:00 +0000
-------------------------------------------

  Change: 6c4477b9e15d5d3213873011b000182ec24fcb66
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-21 04:58:00 +0000

    v0.0100 

  Change: 6f22456cae51b164955abba8f4e4a803033037d0
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-21 04:55:37 +0000

    updated WebService::BaseClientRole requirement version 

  Change: b278b6c7645ad1c3176b8234aaac466830d1cc0e
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-21 04:54:04 +0000

    added set_topic 

  Change: abdcf3b10db13f729071d94bbf06b3aab8449eec
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-21 04:46:37 +0000

    added methods get_members, add_member, and remove_member 

  Change: 41429daaac2a5fbed432ebb727494a79eef848bf
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-21 04:20:41 +0000

    added methods create_room, update_room, and delete_room 

  Change: d6ca762f21bd9b7f99262c5183d88857dfaef3e8
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-07-21 03:34:47 +0000


lib/WebService/HipChat.pm
package WebService::HipChat;
use Moo;
with 'WebService::Client';

our $VERSION = '0.0601'; # VERSION

use Carp qw(croak);
use MIME::Entity;
use JSON qw(encode_json);

has auth_token => ( is => 'ro', required => 1 );

has '+base_url' => ( default => 'https://api.hipchat.com/v2' );

sub BUILD {
    my ($self) = @_;
    $self->ua->default_header(Authorization => "Bearer " . $self->auth_token);
}

sub get_rooms {
    my ($self, %args) = @_;
    return $self->get("/room", $args{query} || {});
}

sub get_room {
    my ($self, $room) = @_;
    croak '$room is required' unless $room;
    return $self->get("/room/$room");
}

sub create_room {
    my ($self, $data) = @_;
    croak '$data is required' unless 'HASH' eq ref $data;
    return $self->post("/room", $data);
}

sub update_room {
    my ($self, $room, $data) = @_;
    croak '$room is required' unless $room;
    croak '$data is required' unless 'HASH' eq ref $data;
    return $self->put("/room/$room", $data);
}

sub set_topic {
    my ($self, $room, $topic) = @_;
    croak '$room is required' unless $room;
    croak '$topic is required' unless $topic;
    if ( 'HASH' eq ref $topic ) {
        $topic = $topic->{topic} or croak '$topic is required';
    }
    return $self->put("/room/$room/topic", { topic => $topic });
}

sub delete_room {
    my ($self, $room) = @_;
    croak '$room is required' unless $room;
    return $self->delete("/room/$room");
}

sub send_notification {
    my ($self, $room, $data) = @_;
    croak '$room is required' unless $room;
    croak '$data is required' unless 'HASH' eq ref $data;
    return $self->post("/room/$room/notification", $data);
}

sub get_webhooks {
    my ($self, $room, %args) = @_;
    croak '$room is required' unless $room;
    return $self->get("/room/$room/webhook", $args{query} || {});
}

sub create_webhook {
    my ($self, $room, $data) = @_;
    croak '$room is required' unless $room;
    croak '$data is required' unless 'HASH' eq ref $data;
    return $self->post("/room/$room/webhook", $data);
}

sub get_members {
    my ($self, $room, %args) = @_;
    croak '$room is required' unless $room;
    return $self->get("/room/$room/member", $args{query} || {});
}

sub add_member {
    my ($self, $room, $user) = @_;
    croak '$room is required' unless $room;
    croak '$user is required' unless $user;
    return $self->put("/room/$room/member/$user", {});
}

sub remove_member {
    my ($self, $room, $user) = @_;
    croak '$room is required' unless $room;
    croak '$user is required' unless $user;
    return $self->delete("/room/$room/member/$user");
}

sub get_users {
    my ($self, %args) = @_;
    return $self->get("/user", $args{query} || {});
}

sub get_user {
    my ($self, $user) = @_;
    croak '$user is required' unless $user;
    return $self->get("/user/$user");
}

sub send_private_msg {
    my ($self, $user, $data) = @_;
    croak '$user is required' unless $user;
    croak '$data is required' unless 'HASH' eq ref $data;
    return $self->post("/user/$user/message", $data);
}

sub get_emoticons {
    my ($self, %args) = @_;
    return $self->get("/emoticon", $args{query} || {});
}

sub get_emoticon {
    my ($self, $emoticon) = @_;
    croak '$emoticon is required' unless $emoticon;
    return $self->get("/emoticon/$emoticon");
}

sub get_room_history {
    my ($self, $room, %args) = @_;
    croak '$room is required' unless $room;
    return $self->get("/room/$room/history/latest", $args{query} || {});
}

sub share_link {
    my ($self, $room, $data) = @_;
    croak '$room is required' unless $room;
    croak '$data is required' unless 'HASH' eq ref $data;
    return $self->post("/room/$room/share/link", $data);
}

sub next {
    my ($self, $data) = @_;
    croak '$data is required' unless 'HASH' eq ref $data;
    my $next = $data->{links}{next} or return undef;
    return $self->get($next);
}

sub share_file {
    my ($self, $room, $data) = @_;
    croak '$room is required' unless $room;
    croak '$data is required' unless 'HASH' eq ref $data;

    my $path = $self->_url("/room/$room/share/file");

    my $msg       = $data->{message};
    my $file      = $data->{file};

    if (! -f $file) {
        warn "File '$file' doesn't exist\n";
        return;
    }

    my $boundary = 'boundary1234567890';

    my $Mime = MIME::Entity->build(
        Type     => 'multipart/related',
        Boundary => $boundary,
    );

    if ( $msg ) {
        my $msg_json = encode_json({ message => $msg });
        $Mime->attach(
            Type     => 'application/json',
            Encoding => '7bit',
            Data     => $msg_json,
        );
    }

    $Mime->attach(
        Path        => $file,
        Disposition => 'attachment; name="file"'
    );

    $Mime->make_multipart();

    return $self->post("/room/$room/share/file",
        $Mime->stringify_body(),
        headers => {
            'content_type' => "multipart/related; boundary=\"$boundary\"",
        },
    );
}


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

WebService::HipChat

=head1 VERSION

version 0.0601

=head1 SYNOPSIS

    my $hc = WebService::HipChat->new(auth_token => 'abc');
    $hc->send_notification('Room42', { color => 'green', message => 'allo' });

    # get paged results:
    my $res = $hc->get_emoticons;
    my @emoticons = @{ $res->{items} };
    while ($res = $hc->next($res)) {
        push @emoticons, @{ $res->{items} };
    }

=head1 DESCRIPTION

This module provides bindings for the
L<HipChat API v2|https://www.hipchat.com/docs/apiv2>.
It also provides the command line utility L<hipchat-send>.

=head1 METHODS

All methods return a hashref.
The C<$room> param can be the id or name of the room.
The C<$user> param can be the id, email address, or @mention name of the user.
If a resource does not exist for the given parameters, undef is returned.

=head2 get_rooms

    get_rooms()
    get_rooms(query => { 'start-index' => 0, 'max-results' => 100 });

Example response:

    {
      items => [
        {
          id => 2,
          links => {
            self => "https://hipchat.com/v2/room/2",
            webhooks => "https://hipchat.com/v2/room/2/webhook",
          },
          name => "General Discussion",
        },
        {
          id => 3,
          links => {
            self => "https://hipchat.com/v2/room/3",
            webhooks => "https://hipchat.com/v2/room/3/webhook",
          },
          name => "Important Stuff",
        },
      links => { self => "https://hipchat.com/v2/room" },
      maxResults => 100,
      startIndex => 0,
    }

=head2 get_room

    get_room($room)

Example response:

    {
      created => "2014-06-25T02:28:04",
      guest_access_url => undef,
      id => 2,
      is_archived => 0,
      is_guest_accessible 0,
      last_active => "2014-07-19T02:40:55+00:00",
      links => {
        self => "https://hipchat.com/v2/room/2",
        webhooks => "https://hipchat.com/v2/room/2/webhook",
      },
      name => "General Discussion",
      owner => {
        id => 1,
        links => { self => "https://hipchat.com/v2/user/1" },
        mention_name => "bob",
        name => "Bob Williams",
      },
      participants => [],
      privacy => "public",
      statistics => {
        links => { self => "https://hipchat.com/v2/room/2/statistics" },
      },
      topic => "hipchat commands",
      xmpp_jid => "1_general_discussion\@conf.btf.hipchat.com",
    }

=head2 create_room

    create_room({ name => 'monkeys' })

Example response:

    {
      id => 46,
      links => { self => "https://hipchat.com/v2/room/46" },
    }

=head2 update_room

    update_room($room, {
        is_archived         => JSON::false,
        is_guest_accessible => JSON::false,
        name                => "Jokes",
        owner               => { id => 17 },
        privacy             => "public",
        topic               => "funny jokes",
    });

=head2 set_topic

    set_topic($room, 'new topic');

=head2 delete_room

    delete_room($room)

=head2 send_notification

    send_notification($room, { color => 'green', message => 'allo' });

=head2 get_webhooks

    get_webhooks($room)
    get_webhooks($room, query => { 'start-index' => 0, 'max-results' => 100 });

Example response:

    {
      items => [
        {
          event => "room_message",
          id => 1,
          links => { self => "https://hipchat.com/v2/room/API/webhook/1" },
          name => "hook1",
          pattern => undef,
          url => "http://yourdomain.org/hipchat-webhook",
        },
      ],
      links => { self => "https://hipchat.com/v2/room/API/webhook" },
      maxResults => 100,
      startIndex => 0,
    }

=head2 create_webhook

    create_webhook($room, {
        url   => 'http://yourdomain.org/hipchat-webhook'
        event => 'room_message',
        name  => 'hook1',
    });

=head2 send_private_msg

    send_private_msg($user, { message => 'allo' });

=head2 get_members

    get_members($room);
    get_members($room, query => { 'start-index' => 0, 'max-results' => 100 });

Example response:

    {
      items => [
        {
          id => 73,
          links => { self => "https://hipchat.com/v2/user/73" },
          mention_name => "momma",
          name => "Yo Momma",
        },
        {
          id => 23,
          links => { self => "https://hipchat.com/v2/user/23" },
          mention_name => "jackie",
          name => "Jackie Chan",
        },
      ],
      links => { self => "https://hipchat.com/v2/room/Test/member" },
      maxResults => 100,
      startIndex => 0,
    }

=head2 add_member

Adds a user to a room.

    add_member($room, $user);

=head2 remove_member

Removes a user from a room.

    remove_member($room, $user);

=head2 get_users

    get_users()
    get_users(query => { 'start-index' => 0, 'max-results' => 100 });

Example response:

    {
      items => [
        {
          id => 1,
          links => { self => "https://hipchat.com/v2/user/1" },
          mention_name => "magoo",
          name => "Matt Wondercookie",
        },
        {
          id => 3,
          links => { self => "https://hipchat.com/v2/user/3" },
          mention_name => "racer",
          name => "Brian Wilson",
        },
      ],
      links => { self => "https://hipchat.com/v2/user" },
      maxResults => 100,
      startIndex => 0,
    }

=head2 get_user

    get_user($user)

Example response:

    {
      created        => "2014-06-20T03:00:28",
      email          => 'matt@foo.com',
      group          => {
                          id => 1,
                          links => { self => "https://hipchat.com/v2/group/1" },
                          name => "Everyone",
                        },
      id             => 1,
      is_deleted     => 0,
      is_group_admin => 1,
      is_guest       => 0,
      last_active    => 1405718128,
      links          => { self => "https://hipchat.com/v2/user/1" },
      mention_name   => "magoo",
      name           => "Matt Wondercookie",
      photo_url      => "https://hipchat.com/files/photos/1/abc.jpg",
      presence       => {
                          client => {
                            type => "http://hipchat.com/client/linux",
                            version => 98,
                          },
                          idle => 3853,
                          is_online => 1,
                          show => "away",
                        },
      timezone       => "America/New_York",
      title          => "Hacker",
      xmpp_jid       => '1_1@chat.hipchat.com',
    }

=head2 get_emoticons

    get_emoticons()
    get_emoticons(query => { 'start-index' => 0, 'max-results' => 100 });

Example response:

    {
      items => [
        {
          id => 166,
          links => { self => "https://hipchat.com/v2/emoticon/166" },
          shortcut => "dog",
          url => "https://hipchat.com/files/img/emoticons/1/dog.png",
        },
      ],
      links => { self => "https://hipchat.com/v2/emoticon" },
      maxResults => 100,
      startIndex => 0,
    }

=head2 get_emoticon

    get_emoticon()

Example response:

    {
      creator => {
        id => 11,
        links => { self => "https://hipchat.com/v2/user/11" },
        mention_name => "bob",
        name => "Bob Ray",
      },
      height => 30,
      id => 203,
      links => { self => "https://hipchat.com/v2/emoticon/203" },
      shortcut => "dog",
      url => "https://hipchat.com/files/img/emoticons/1/dog.png",
      width => 30,
    }

=head2 get_room_history

    $hc->get_room_history($room)
    $hc->get_room_history($room, { 'max-results' => 5 });

Example response:

   {
    items        [
        [0] {
            date       "2014-11-13T10:48:33.322506+00:00",
            from       {
                id             123456,
                links          {
                    self   "https://api.hipchat.com/v2/user/123456"
                },
                mention_name   "Bob",
                name           "Bob Williams"
            },
            id         "38988c8c-9120-44ce-87c5-6731a7a3b6",
            mentions   [],
            message    "heres a message and a link http://www.sun.com/",
            type       "message"
        },
        [1] {
            date       "2014-11-13T10:49:02.377436+00:00",
            from       {
                id             123456,
                links          {
                    self   "https://api.hipchat.com/v2/user/123456"
                },
                mention_name   "Bob",
                name           "Bob Williams"
            },
            id         "c1f47537-6506-4f46-b820-eaade5adc5",
            mentions   [],
            message    "A message",
            type       "message"
        }
    ],
    links        {
        self   "https://api.hipchat.com/v2/room/XXX/history/latest"
    },
    maxResults   2,
    startIndex   0
   }

=head2 share_link

    $hc->share_link($room, { message => 'msg', link => 'http://www.sun.com' });

=head2 share_file

    $hc->share_file($room, { message => 'msg', file => '/tmp/file.png' });

=head2 next

    next($data)

Returns the next page of data for paginated responses.

Example:

    my $res = $hc->get_emoticons;
    my @emoticons = @{ $res->{items} };
    while ($res = $hc->next($res)) {
        push @emoticons, @{ $res->{items} };
    }

=head1 CONTRIBUTORS

=over

=item Chris Hughes <chris@lokku.com>

=back

=head1 AUTHOR

Naveed Massjouni <naveed@vt.edu>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Naveed Massjouni.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
